portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -44,31 +56,101 @@
+      json.addProperty("name", name);
+    } else {
+      // Add a login link to the response if the user is not logged in.
+      json.addProperty("loginUrl", userService.createLoginURL(redirectUrl));
+    }
 
+    return json;
+  }
+
+  /**
+   * Gets a list of all the comments in JSON format translated to the selected language.
+   */
+  private JsonElement queryComments(HttpServletRequest request) {
     // Query up to maxComments comment entities from Datastore with the user's specified sorting option.
+    int maxComments = getMaxCommentsToReturn(request);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     Query query = createCommentQuery(request);
-    List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(maxComments));
+    List<Entity> entities = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(maxComments));
 
-    // Construct a list of comments from the queried entities.
-    List<Comment> comments = results.stream().map((entity) -> {
-      long id = entity.getKey().getId();
-      String author = (String) entity.getProperty("author");
-      String commentText = (String) entity.getProperty("commentText");
-      long timestamp = (long) entity.getProperty("timestamp");
+    // Check if there are comments to translate.
+    if (entities.isEmpty()) {
+      return new JsonArray();
+    } else {
+      // Construct a stream of comment texts from the queried entities.
+      Stream<String> commentTexts = 
+          entities.stream().map((entity) -> (String) entity.getProperty("commentText"));
 
-      return new Comment(id, author, commentText, timestamp);
-    }).collect(Collectors.toList());
+      // Translate the comments to the selected language, preserving order.
+      String languageCode = request.getParameter("language-code");
+      Stream<String> translatedCommentTexts = translateComments(commentTexts, languageCode);
 
-    // Convert the list of comments to JSON.
-    String json = convertToJson(comments);
+      List<Comment> comments = createComments(entities.stream(), translatedCommentTexts);     
+      return convertToJsonElement(comments);
+    }
+  }
 
-    // Send the JSON as the response.
-    response.setContentType("text/html;");
-    response.getWriter().println(json);
+  /** 
+   * Returns the name of the user with id, or null if a name has not been set. 
+   */
+  private Optional<String> getUserName(String id) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query =
+        new Query("UserInfo")
+            .setFilter(new Query.FilterPredicate("id", Query.FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return Optional.empty();
+    }
+
+    String name = (String) entity.getProperty("name");
+    return Optional.of(name);
```

Comment:
You can shorten this to single line using `Optional.ofNullable()` [1] with `.map` [2]

```
return Optional.ofNullable(entity).map(...)
```

[1] https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#ofNullable-T-
[2] https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#map-java.util.function.Function-